elfpy/utils/apeworx_integrations.py

```diff
@@ -0,0 +1,250 @@
+    hyperdrive_contract: ape.contracts.base.ContractInstance
+        Ape project `ContractInstance
+        <https://docs.apeworx.io/ape/stable/methoddocs/contracts.html#ape.contracts.base.ContractInstance>`_
+        wrapped around the initialized MockHyperdriveTestnet smart contract.
+    agent_address: ape.api.accounts.AccountAPI
+        Ape address container, or `AccountAPI
+        <https://docs.apeworx.io/ape/stable/methoddocs/api.html#ape.api.accounts.AccountAPI>`_
+        representing the agent which is executing the action.
+    bond_amount: int
+        Unsigned int-256 representation of the bond amount that the agent wishes to sell.
+    maturity_time: int
+        Unsigned int-256 representation of the maturity time of the short being sold.
+        This is equal to the pool position duration plus the checkpoint time that is
+        closest to (but before) the corresponding open trade.
+
+    Returns
+    -------
+    Tuple[dict[str, Any], ape_ethereum.transactions.Receipt]
+        A tuple containing an update dictionary for the Hyperdrive pool state
+        as well as the Ape-ethereum transaction `receipt
+        <https://docs.apeworx.io/ape/stable/methoddocs/api.html#ape.api.transactions.ReceiptAPI>`_.
+    """
+    # pylint: disable=too-many-locals
+    with ape.accounts.use_sender(agent_address):  # sender for contract calls
+        # Ensure requested close amount is not greater than what is available in the pool
+        trade_asset_id = encode_asset_id(trade_prefix, maturity_time)
+        agent_balance = hyperdrive_contract.balanceOf(trade_asset_id, agent_address)  # type: ignore
+        if bond_amount < agent_balance:
+            trade_bond_amount = bond_amount
+        else:
+            trade_bond_amount = agent_balance
+            logging.warning(
+                "bond_amount=%g is greater than or equal to the Hyperdrive pool balance=%g",
+                bond_amount,
+                trade_bond_amount,
+            )
+        # Close the position
+        min_output = 0
+        as_underlying = False  # mockHyperdriveTestNet does not support as_underlying=True
+        if trade_prefix == AssetIdPrefix.LONG:
+            tx_receipt = hyperdrive_contract.closeLong(  # type: ignore
+                maturity_time,
+                trade_bond_amount,
+                min_output,
+                agent_address,
+                as_underlying,
+            )
+        elif trade_prefix == AssetIdPrefix.SHORT:
+            tx_receipt = hyperdrive_contract.closeShort(  # type: ignore
+                maturity_time,
+                trade_bond_amount,
+                min_output,
+                agent_address,
+                as_underlying,
+            )
+        else:
+            raise ValueError(f"{trade_prefix=} must be 0 (long) or 1 (short).")
+        # Return the updated pool state & transaction result
+        transfer_single_event = get_transaction_trade_event(tx_receipt)
```

Comment:
oh i see, ok maybe we rename it to just `get_transfer_single_event` and then we can use it wherever (like here) where we know that TransferSingle has what we want